duplicate_url_discarder/url_canonicalizer.py

```diff
@@ -13,11 +15,27 @@
 
 
 class UrlCanonicalizer:
-    def __init__(self, rule_paths: Iterable[Union[str, os.PathLike]]) -> None:
+    def __init__(self) -> None:
+        self.url_matcher = URLMatcher()
+        self.processors: Dict[int, UrlProcessorBase] = {}
+
+    @staticmethod
+    def _is_url(path: str) -> bool:
+        return path.startswith("http://") or path.startswith("https://")
+
+    async def load_rules(self, rule_paths: Iterable[Union[str, os.PathLike]]) -> None:
+        if self.processors:
+            raise RuntimeError("UrlCanonicalizer.load_rules() can only be called once.")
+
         rules: Set[UrlRule] = set()
         full_rule_count = 0
         for rule_path in rule_paths:
-            data = Path(rule_path).read_text()
+            data: str
+            if isinstance(rule_path, str) and self._is_url(rule_path):
+                response = await maybe_deferred_to_future(treq.get(rule_path))
+                data = await response.text()
```

Comment:
Let's handle the case when the rules were not successfully retrieved due to things like connection issues, timeouts, etc:
1. Logging the error
2. Terminate the crawl - I think this would be a good behavior since if the spider proceeds without any rules, the user would accumulate a lot of requests due to unfiltered requests.